src/dtx/dtx_rpc.c

```diff
@@ -1050,11 +1050,31 @@ dtx_refresh_internal(struct ds_cont_child *cont, int *check_count,
 			/*
 			 * The leader does not have related DTX info, we may miss
 			 * related DTX abort request, so let's abort it locally.
+			 *
+			 * NOTE:
+			 * There is race between DTX refresh RPC being triggered on current engine
+			 * and DTX commit RPC on remote leader. Related DTX entry on current engine
+			 * may has been committed by race before DTX refresh RPC being replied. And
+			 * it is possible that related DTX entry on remote leader may be removed by
+			 * DTX aggregation before the DTX refresh RPC being handled on the leader.
+			 * Under such case, the leader will reply -DER_NONEXIST to the DTX refresh
+			 * RPC sponsor. Let's check such case to avoid confused abort failure.
 			 */
-			rc1 = vos_dtx_abort(cont->sc_hdl, &dsp->dsp_xid, dsp->dsp_epoch);
-			if (rc1 == 0 || rc1 == -DER_NONEXIST || rc1 == -DER_NO_PERM || !failout) {
+
+			rc1 = vos_dtx_check(cont->sc_hdl, &dsp->dsp_xid,
+					    NULL, NULL, NULL, NULL, false);
+			if (rc1 == DTX_ST_COMMITTED || rc1 == DTX_ST_COMMITTABLE ||
+			    rc1 == -DER_NONEXIST) {
```

Comment:
just confirm that for the case of "rc1 == -DER_NONEXIST", can it just ignore it, or should it "reply -DER_NONEXIST to the DTX refresh RPC sponsor" as the comment, seems the rc1 here will be ignored?

Reply:
The caller of current "vos_dtx_check" is the DTX refresh RPC sponsor. We call it just because the DTX leader replied "-DER_NONEXIST" to the DTX refresh RPC sponsor. We need to check the DTX status before we abort it.

Reply:
I'm confused, when leader replied -DER_NONEXIST, does vos_dtx_check() also -DER_NONEXIST or it should return DER_COMMITTED.